mleap-spark-testkit/src/main/scala/org/apache/spark/ml/parity/SparkParityBase.scala

```diff
@@ -111,31 +111,28 @@ abstract class SparkParityBase extends FunSpec with BeforeAndAfterAll {
   }
 
   def assertModelTypesMatchTransformerTypes(model: Model, shape: NodeShape, exec: UserDefinedFunction): Unit = {
-    val modelInputTypes = shape.inputs.
-      map(_._2.port).
-      map(n => model.inputSchema.getField(n).get.dataType).
-      toSeq
+    val inputFields = model.inputSchema.fields.map(_.name)
+    val modelInputTypes = model.inputSchema.fields.map(_.dataType)
     val transformerInputTypes = exec.inputs.flatMap(_.dataTypes)
 
-    val modelOutputTypes = shape.outputs.
-      map(_._2.port).
-      map(n => model.outputSchema.getField(n).get.dataType).
-      toSeq
```

Comment:
This change (and the one for the inputs) is probably the most contentious thing in this PR.  I /think/ it shouldn't be doing this remapping since the `exec.outputTypes` are not remapped.  For the xgboost models, it was encountering an issue where the schemas were mis-aligned.  But when I remove this remapping, then they are aligned as expected.  This check is between the transformer's udf and the model's ouput schema, so the node shape mapping shouldn't have happened yet.

Please let me know if I'm misunderstanding something though 🙂

Reply:
This test isn't actually checking the Transformer's schema though.  It checks the Transformer's udf vs the Transformer's model's schema.  Both of these things sit "behind" the NodeShape remapping and are supposed to match exactly.

I.e., Transformer internal logic is roughly `return udf(inputs => model(inputs))` and this test is assuring that the udf input-/output-schema matches `model.apply`'s input-/output- schema.  Those things must always match exactly otherwise the udf/model won't align.  The current test was reordering the model schema half of the equation via the NodeShape remapping -- which could change the order of the fields in ways that we can't control.